discovery-provider/src/queries/get_premium_track_signatures.py

```diff
@@ -218,19 +227,167 @@ def _get_eth_nft_gated_track_signatures(
+            bytes(METADATA_PROGRAM_ID_PK),
+            bytes(PublicKey(mint_address)),  # type: ignore
+        ],
+        METADATA_PROGRAM_ID_PK,
+    )[0]
+
+
+def _does_user_own_sol_nft_collection(
+    collection_mint_address: str, user_sol_wallets: List[str]
+):
+    if not solana_client_manager:
+        return False
+
+    for wallet in user_sol_wallets:
+        try:
+            result = solana_client_manager.get_token_accounts_by_owner(wallet)
+            token_accounts = result["value"]
+            nft_mints = list(
+                map(
+                    lambda item: item["account"]["data"]["parsed"]["info"]["mint"],
+                    filter(
+                        lambda item: item["account"]["data"]["parsed"]["info"][
+                            "tokenAmount"
+                        ]["amount"]
+                        != "0"
+                        and item["account"]["data"]["parsed"]["info"]["tokenAmount"][
+                            "decimals"
+                        ]
+                        == 0,
+                        token_accounts,
+                    ),
+                )
+            )
+            metadata_accounts = list(map(_get_metadata_account, nft_mints))
+            datas = list(
+                map(
+                    lambda metadata_account: base64.b64decode(
+                        solana_client_manager.get_account_info(metadata_account)[
+                            "value"
+                        ]["data"][0]
+                    ),
+                    metadata_accounts,
+                )
+            )
+            metadatas = list(map(_unpack_metadata_account, datas))
+            collections = list(
+                filter(
+                    lambda collection: collection
+                    and collection["verified"]
+                    and collection["key"].decode() == collection_mint_address,
+                    list(map(lambda metadata: metadata["collection"], metadatas)),
+                )
+            )
+            if collections:
+                return True
+        except Exception as e:
+            logger.error(
+                f"Could not get nft balance for nft collection {collection_mint_address} and user wallet {wallet}. Error: {e}"
+            )
+    return False
+
+
 def _get_sol_nft_gated_track_signatures(
     user_wallet: str,
     sol_associated_wallets: List[str],
     tracks: List[Track],
 ):
-    return {}
+    track_signature_map = {}
+
+    # Build a map of collection mint address -> track ids
+    # so that only one chain call will be made for premium tracks
+    # that share the same nft collection gate.
+    collection_track_map = defaultdict(list)
+    for track in tracks:
+        collection_mint_address = track.premium_conditions["nft_collection"]["address"]  # type: ignore
+        collection_track_map[collection_mint_address].append(track.track_cid)
+
+    with concurrent.futures.ThreadPoolExecutor() as executor:
```

Comment:
I'm sorry to do this to you, but let's write this with asyncio. It's more modern, maintainable, and better perf. LMK if codepointer would be helpful

Reply:
we now use asyncio for the calls to `get_account_info` which improved perf, but using asyncio in this part here did not yield better perf; actually made it a bit slower...
going to merge this as-is now with the asyncio win above.

can revisit this later if you think there is a way to change this part and make it better too